examples/pennylane_run_quantum_natural_gradient.py

```diff
@@ -0,0 +1,451 @@
+
+@qml.qnode(dev)
+def circuit(params):
+    # non-parametrized gates
+    qml.RY(np.pi/4, wires=0)
+    qml.RY(np.pi/3, wires=1)
+    qml.RY(np.pi/7, wires=2)
+
+    # Parametrized layer 1
+    qml.RZ(params[0], wires=0)
+    qml.RZ(params[1], wires=1)
+
+    # non-parametrized gates
+    qml.CNOT(wires=[0, 1])
+    qml.CNOT(wires=[1, 2])
+
+    # Parametrized layer 2
+    qml.RY(params[2], wires=1)
+    qml.RX(params[3], wires=2)
+
+    # non-parametrized gates
+    qml.CNOT(wires=[0, 1])
+    qml.CNOT(wires=[1, 2])
+
+    return qml.expval(qml.PauliY(0))
+
+params = np.array([0.432, -0.123, 0.543, 0.233])
+
+##############################################################################
+# The above circuit consists of 4 parameters, with two distinct parametrized
+# layers of 2 parameters each.
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng1.png
+#     :align: center
+#     :width: 90%
+#     :target: javascript:void(0)
+#
+# |
+#
+# Therefore, the block-diagonal approximation consists of two
+# :math:`2\times 2` matrices, :math:`g^{(1)}` and :math:`g^{(2)}`.
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng2.png
+#     :align: center
+#     :width: 30%
+#     :target: javascript:void(0)
+#
+# To compute the first block-diagonal :math:`g^{(1)}`, we create subcircuits consisting
+# of all gates prior to the layer, and observables corresponding to
+# the *generators* of the gates in the layer:
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng3.png
+#     :align: center
+#     :width: 30%
+#     :target: javascript:void(0)
+
+g1 = np.zeros([2, 2])
+
+def layer1_subcircuit(params):
+    """This function contains all gates that
+    precede parametrized layer 1"""
+    qml.RY(np.pi/4, wires=0)
+    qml.RY(np.pi/3, wires=1)
+    qml.RY(np.pi/7, wires=2)
+
+##############################################################################
+# We then post-process the measurement results in order to determine :math:`g^{(1)}`,
+# as follows.
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng4.png
+#     :align: center
+#     :width: 50%
+#     :target: javascript:void(0)
+#
+# We can see that the diagonal terms are simply given by the variance:
+
+@qml.qnode(dev)
+def layer1_diag(params):
+    layer1_subcircuit(params)
+    return var(qml.PauliZ(0)), var(qml.PauliZ(1))
+
+# calculate the diagonal terms
+varK0, varK1 = layer1_diag(params)
+g1[0, 0] = varK0/4
+g1[1, 1] = varK1/4
+
+##############################################################################
+# The following two subcircuits are then used to calculate the
+# off-diagonal covariance terms of :math:`g^{(1)}`:
+
+@qml.qnode(dev)
+def layer1_off_diag_single(params):
+    layer1_subcircuit(params)
+    return expval(qml.PauliZ(0)), expval(qml.PauliZ(1))
+
+@qml.qnode(dev)
+def layer1_off_diag_double(params):
+    layer1_subcircuit(params)
+    ZZ = np.kron(np.diag([1, -1]), np.diag([1, -1]))

```

Comment:
Will need to be upgraded soon, once #267 is fully merged

Reply:
Done!